Access `buildConfig` in a configuration-cache-compatible way

I'm trying to follow the instructions to [generate constants from resources files](https://github.com/gmazzo/gradle-buildconfig-plugin?tab=readme-ov-file#generate-constants-from-resources-files). However, the line

https://github.com/gmazzo/gradle-buildconfig-plugin/blob/5fd8a3276a061fa26f104ddf5c3813f2d2bd4d2e/README.md?plain=1#L260

seems to cause

```
FAILURE: Build failed with an exception.

* What went wrong:
Configuration cache problems found in this build.

1 problem was found storing the configuration cache.
- Task `:plugins:package-managers:swiftpm-package-manager:generateResourcesConstants` of type `org.gradle.api.DefaultTask`: cannot serialize Gradle script object references as these are not supported with the configuration cache.
  See https://docs.gradle.org/8.14.2/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
```

Is there any guidance on how to do this in a configuration-cache-compatible way?

--- a/README.md
+++ b/README.md
@@ -257,22 +257,16 @@ myproject
 If you add in your `build.gradle.kts`:
 
 ```kotlin
-val buildResources = buildConfig.forClass("BuildResources")
-val generateResourcesConstants by tasks.registering {
-    val resources = sourceSets["main"].resources.asFileTree
+buildConfig.forClass("BuildResources") {
+  buildConfigField("A_CONSTANT", "aConstant")
 
-    inputs.files(resources)
-    doFirst {
-        resources.visit(Action<FileVisitDetails> {
-            val name = path.toUpperCase().replace("\\W".toRegex(), "_")
+  sourceSets["main"].resources.asFileTree.visit {
+    if (!isDirectory) {
+      val name = path.uppercase().replace("\\W".toRegex(), "_")
 
-            buildResources.buildConfigField("java.io.File", name, "File(\"$path\")")
-        })
+      buildConfigField(name, File(path))
     }
-}
-
-tasks.generateBuildConfig {
-    dependsOn(generateResourcesConstants)
+  }
 }
 ```
 

--- a/demo-project/groovy/build.gradle
+++ b/demo-project/groovy/build.gradle
@@ -208,17 +208,11 @@ sourceSets {
 buildConfig.forClass("BuildResources") {
     buildConfigField('String', 'A_CONSTANT', '"aConstant"')
 
-    def files = sourceSets.main.resources.sourceDirectories.asFileTree
-    buildConfigFields.addAllLater(provider {
-        List<BuildConfigField> fields = []
-        files.asFileTree.visit {
+    sourceSets.main.resources.asFileTree.visit {
+        if (!it.directory) {
             def name = it.path.toUpperCase().replaceAll("\\W", "_")
 
-            fields.add(objects.newInstance(BuildConfigField, name).with { _ ->
-                type(File)
-                expression("new File(\"${it.path}\")")
-            })
+            buildConfigField(File.class, name, new File(it.path))
         }
-        return fields
-    })
+    }
 }

--- a/demo-project/kts/build.gradle.kts
+++ b/demo-project/kts/build.gradle.kts
@@ -217,28 +217,7 @@ val versionsSS = buildConfig.sourceSets.register("Versions") {
     buildConfigField("myDependencyVersion", "1.0.1")
 }
 
-/**
- *  A task that iterates over your classpath resources and generate constants for them
- */
-buildConfig.forClass("BuildResources") {
-    buildConfigField("A_CONSTANT", "aConstant")
-
-    val files = sourceSets["main"].resources.sourceDirectories.asFileTree
-    buildConfigFields.addAllLater(provider {
-        val fields = mutableListOf<BuildConfigField>()
-        files.visit {
-            val name = path.uppercase().replace("\\W".toRegex(), "_")
-
-            fields.add(objects.newInstance<BuildConfigField>(name).apply {
-                type(File::class)
-                expression("File(\"$path\")")
-            })
-        }
-        fields
-    })
-}
-
-// example of a custom generator that builds into XML
+// Example: Custom generator that builds into XML
 val propertiesSS = buildConfig.sourceSets.register("properties") {
     buildConfigField("value1", "AAA")
     buildConfigField("value2", "BBB")
@@ -261,6 +240,19 @@ val propertiesSS = buildConfig.sourceSets.register("properties") {
         }
 
     })
+}
+
+// Example: Generate constants from resources files
+buildConfig.forClass("BuildResources") {
+    buildConfigField("A_CONSTANT", "aConstant")
+
+    sourceSets["main"].resources.asFileTree.visit {
+        if (!isDirectory) {
+            val name = path.uppercase().replace("\\W".toRegex(), "_")
+
+            buildConfigField(name, File(path))
+        }
+    }
 }
 
 sourceSets.main {

